engine/src/graphics/device.cpp

```diff
@@ -132,13 +133,80 @@ namespace gen
 			queueCreateInfos.push_back(queueCreateInfo);
 		}
 
-		vk::DeviceCreateInfo const createInfo({}, static_cast<u32>(queueCreateInfos.size()), queueCreateInfos.data());
+		vk::DeviceCreateInfo const createInfo({}, static_cast<u32>(queueCreateInfos.size()), queueCreateInfos.data(),
+											  0,	   // ignored by spec
+											  nullptr, // ignored by spec
+											  static_cast<u32>(deviceExtensions.size()), deviceExtensions.data(),
+											  nullptr // to be used later
+		);
 
 		m_device = m_gpu.physicalDevice.createDeviceUnique(createInfo);
 
+		// display all device extensions using m_device
+		auto availableExtensions = m_gpu.physicalDevice.enumerateDeviceExtensionProperties();
+		auto aExt				 = std::stringstream{"vulkan"};
+		for (std::size_t i = 0; i < availableExtensions.size(); i++)
+		{
+			aExt << "\tExtension [" << i << "]"
+				 << " | Name: " << availableExtensions[i].extensionName << " | Spec Ver: " << availableExtensions[i].specVersion << "\n";
+		}
+		m_logger.debug("Found Device Extensions: \n{}\n", aExt.str());
+
 		m_graphicsQueue = m_device->getQueue(m_gpu.queueFamily, 0);
 	}
 
+	void GraphicsDevice::createSwapChain(const Window & window)
+	{
+		std::vector<vk::SurfaceFormatKHR> const formats = m_gpu.physicalDevice.getSurfaceFormatsKHR(m_surface.get());
+		assert(!formats.empty());
+
+		// If the format of the first element in the 'formats' array is undefined,
+		// use the RGBA 8-bit per channel normalized format (eB8G8R8A8Unorm) as the default.
+		// Otherwise, use the format that is already specified in the first element.
+		vk::Format const format = (formats[0].format == vk::Format::eUndefined) ? vk::Format::eB8G8R8A8Unorm : formats[0].format;
+
+		vk::SurfaceCapabilitiesKHR const surfaceCapabilities = m_gpu.physicalDevice.getSurfaceCapabilitiesKHR(m_surface.get());
+
+		vk::Extent2D swapchainExtent{};
+		if (surfaceCapabilities.currentExtent.width == std::numeric_limits<u32>::max())
+		{
+			// If the surface size is undefined, the size is set to the size of the images requested.
+			swapchainExtent.width =
+				std::clamp(static_cast<u32>(window.getWidth()), surfaceCapabilities.minImageExtent.width, surfaceCapabilities.maxImageExtent.width);
```

Comment:
Does `getWidth()` return window width or framebuffer width? We want the latter.

Reply:
Ahh it is returning the windows size not the framebuffer. I can fix that up real quick ^^